use subtest for table units (pkg/scheduler/core)

xchapter7x opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->
**What this PR does / why we need it**: Update scheduler's unit table tests to use subtest

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:
breaks up PR: https://github.com/kubernetes/kubernetes/pull/63281
/ref #63267

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
This PR will leverage subtests on the existing table tests for the scheduler units.
Some refactoring of error/status messages and functions to align with new approach.

```

misterikkit commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -692,13 +746,15 @@ func TestInvalidateCachedPredicateItemOfAllNodes(t *testing.T) {
+				if _, exist := algorithmCache.predicatesCache.Get(testPredicate); exist {
+					t.Errorf("Failed: cached item for predicate key: %v on node: %v should be invalidated",
+						testPredicate, test.nodeName)
+					return
```
This won't exit the loop like the break did previously. I think it should be fine to run to the end of the loop though.

misterikkit commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -692,13 +746,15 @@ func TestInvalidateCachedPredicateItemOfAllNodes(t *testing.T) {
 
 	// there should be no cached predicate any more
 	for _, test := range tests {
```
Taking a closer look at this test, I don't like that the various test cases seem to leak state into each other (i.e. loop over the tests twice with the same ecache.)

It's definitely beyond the scope of this PR, but I just wanted to point it out. I should file a separate issue.

misterikkit commented on pkg/scheduler/core/generic_scheduler_test.go:
```diff
@@ -156,31 +158,33 @@ func TestSelectHost(t *testing.T) {
-				if !test.possibleHosts.Has(got) {
-					t.Errorf("got %s is not in the possible map %v", got, test.possibleHosts)
+		t.Run(test.name, func(t *testing.T) {
+			// increase the randomness
```
More commentary on the pre-existing tests: This isn't super random...

misterikkit commented on pkg/scheduler/core/generic_scheduler_test.go:
```diff
@@ -396,39 +400,41 @@ func TestGenericScheduler(t *testing.T) {
+			machine, err := scheduler.Schedule(test.pod, schedulertesting.FakeNodeLister(makeNodeList(test.nodes)))
+
+			if !reflect.DeepEqual(err, test.wErr) {
+				t.Errorf("Failed : Unexpected error: %v, expected: %v", err, test.wErr)
```
Can also remove the "Failed :" prefix from these messages.

misterikkit commented on pkg/scheduler/core/generic_scheduler_test.go:
```diff
@@ -629,55 +639,61 @@ func TestZeroRequest(t *testing.T) {
-			t.Errorf("unexpected error: %v", err)
-		}
-		for _, hp := range list {
-			if test.test == "test priority of larger pod with machine with zero-request pod" {
```
I don't know what to say about this...

bsalamat replied:
This is really ugly, but it existed. We should consider fixing this in a separate PR.

misterikkit reviewed (commented):
It looks like you pulled some test bodies into helper functions and left others inline. My guess is that you did this to improve readability, but I feel that keeping them all inline is more readable. Even though it's a lot of code, it only makes sense in the context of the one test where it is used.

This is clearly subjective, so I'm open to be convinced otherwise. What is the threshold for a test size before it should be pulled into a helper?

misterikkit commented:
/ok-to-test

xchapter7x commented:
/test pull-kubernetes-e2e-kops-aws

xchapter7x commented:
/test pull-kubernetes-e2e-kops-aws

misterikkit commented:
> It looks like you pulled some test bodies into helper functions and left others inline.
Since this pattern appears in a few of your PRs, I'd like to resolve it first. How do you feel about keeping tests inline?

xchapter7x commented:
@misterikkit , i think the discussion to break out the tests or to keep them inline can be deferred.
Thinking about it now, it feels out of scope to this effort.

So, I agree, lets keep them inline for now. 
let me re-work this PR. 
im gonna put this one back to WIP for now. stay tuned.

xchapter7x committed:
use subtest for table units

remove duplicate testname from error msg

remove subtest for test setup loop

do not break on test failure

  https://github.com/kubernetes/kubernetes/pull/63665#discussion_r203571355

remove duplicate test.name in output

  https://github.com/kubernetes/kubernetes/pull/63665#discussion_r203574001
  https://github.com/kubernetes/kubernetes/pull/63665#discussion_r203574012

xchapter7x commented:
/test pull-kubernetes-e2e-gce

xchapter7x commented:
/test pull-kubernetes-e2e-gce

xchapter7x commented:
/assign @misterikkit
this PR should be in a good spot if you have a moment to take a look. thanks.

misterikkit commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -342,40 +342,42 @@ func TestUpdateResult(t *testing.T) {
+
+			cachedMapItem, ok := ecache.algorithmCache[test.nodeName][test.predicateKey]
+			if !ok {
+				t.Errorf("Failed: %s, can't find expected cache item: %v",
```
Remove test.name  and "Failed" from log messages. I think you had done this in a previous version, and accidentally reverted it?

misterikkit commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -689,31 +697,40 @@ func TestInvalidateCachedPredicateItemOfAllNodes(t *testing.T) {
-			test.cache,
-			node,
-		)
+		t.Run(test.name, func(t *testing.T) {
```
Given the nature of this test, I think we should actually run this first loop in the main test, and use t.Run in the second loop like you are already doing.

My reasoning is that this test uses the same ecache object across all test cases. (I think that should be fixed up in a separate PR.) so this loop feels more like setup than the actual test.

bsalamat replied:
I agree with @misterikkit. You shouldn't `t.Run` in this loop.

xchapter7x commented:
/test pull-kubernetes-e2e-kops-aws

bsalamat commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -708,12 +714,19 @@ func TestInvalidateCachedPredicateItemOfAllNodes(t *testing.T) {
+				if _, exist := algorithmCache[testPredicate]; exist {
+					t.Errorf("Failed: cached item for predicate key: %v on node: %v should be invalidated",
+						testPredicate, test.nodeName)
+					breakOnTestFailure = true
```
I think we can remove this and continue testing other cases when one fails.

bsalamat commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -476,51 +476,53 @@ func TestLookupResult(t *testing.T) {
+			// returned invalid should match expectedInvalidPredicateKey or expectedInvalidEquivalenceHash
+			if test.equivalenceHashForUpdatePredicate != test.equivalenceHashForCalPredicate {
+				if invalid != test.expectedInvalidEquivalenceHash {
+					t.Errorf("Failed: %s, expected invalid: %v, but got: %v",
```
To be consistent, please remove `test.name` here.

bsalamat commented on pkg/scheduler/core/equivalence_cache_test.go:
```diff
@@ -476,51 +476,53 @@ func TestLookupResult(t *testing.T) {
+				}
+			} else {
+				if invalid != test.expectedInvalidPredicateKey {
+					t.Errorf("Failed: %s, expected invalid: %v, but got: %v",
```
ditto

bsalamat reviewed (commented):
Thanks, @xchapter7x!

xchapter7x commented:
@bsalamat thanks for the feedback. 

changes are in and merge conflicts should also be resolved.
Just waiting for all checks to go green.

- John

bsalamat reviewed (approved):
/lgtm

Thanks, @xchapter7x!
